File: CA3 Calculators/Python Calculator/calculator.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jan 20 19:50:22 2018

@author: Admin
"""

class Calculator(object):
    
    def __init__(self):
       self.tempOper2 = ""
       self.tempOper1 = ""
       self.number = ""
       self.total = None
       self.answer = None
    
    isNumber = False
   
            
    operators2Num = {'+': lambda x,y:x+y,
                     '-': lambda x,y:x-y,
                     '/': lambda x,y:x/y if y != 0.0 else 0.0,
                     '*': lambda x,y:x*y,
                     '%': lambda x,y:x%y if y != 0.0 else 0.0,
                     }
    
    operators1Num = {'x^2': lambda x: x**2,
                     'sqR': lambda x: x**0.5 if x >= 0 else 0,
                     'cube': lambda x: x**3,
                     'cubeR': lambda x: abs(x)**(1./3),
                     'abs': lambda x: abs(x),
                     }
    
    def checkIsNumber(self,input):
        #check if user input is a number
        try:
            self.isNumber = float(input)
            return True
        except ValueError:
            return False
    
    def setInput(self,input):
        #assigning the validated input to either a number of operator variable
       if self.checkIsNumber(input) == True:
           #Input number can be concatenated until an operator has been entered
           if self.tempOper1 == "" and self.tempOper2 == "" and self.number == "":
               self.number =  str(input)
               return self.number
           elif self.tempOper1 == "" and self.tempOper2 == "":
               if float(input) < 1 and float(input) > 0:
                   self.number = float(self.number) + float(input)
               else:
                   self.number = str(self.number) + str(input)
                   return self.number
            #Calcs requiring 2 input numbers - second number assignment
           elif self.tempOper2 != "": 
               self.number = str(input)
               return self.number
           return self.number
       #Input is not a number => assigning the validated operator to the variable
       #Assigning the stored number into total
       elif str(input) in self.operators1Num:
           self.tempOper1 = input
           self.total = self.number
           return self.tempOper1
       else:
           self.tempOper2 = input
           self.total = self.number
           self.number = ""      #resetting number to "" for the required second input
           return self.tempOper2
       
       
        
    def getCalculator1NumResult(self,operation):
       self.answer = self.operators1Num[operation](float(self.total))
       self.tempOper1 = ""
       self.total = None
       self.number = ""
       return self.answer
   
    def getCalculator2NumResult(self,operation):
        self.answer = self.operators2Num[operation](float(self.total), float(self.number))
        self.tempOper2 = ""
        self.total = None
        self.number = ""
        return self.answer
        
        
        
        
        
        
   
            
        
        

File: CA3 Calculators/TDDCalculator/TDDCalculator/Calculator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TDDCalculator
{
    
    public class Calculator
    {
        private static readonly string[] operations2Num = { "+", "-", "/", "*", "%" };
        private static readonly string[] operations1Num = { "x^2", "sqR", "cube", "cubeR", "abs" };
        public string Number { get; set; }
        public string Total { get; set; }
        public string tempOper1;
        public string tempOper2;
        public double answer;

        //Methods - assigning the inputs
        public string SetInput(string tempInput)
        {
            //Check if tempInput is a number = assign to Number
            if (IsValidNumber(tempInput))
            {
                if (tempOper1 == null && tempOper2 == null)//concatenate the inputs into 1 number
                {
                    Number = Number + tempInput;
                    return Number;
                }
                else if(tempOper2 != null)//second number input for calcs requiring 2 inputs
                {
                    Number = tempInput;
                    return Number;
                }
                return Number;
            }
            //else assign the valid "operations" to tempOper1 or tempOper2
            else if(IsValidOperation1Num(tempInput))
            {
                tempOper1 = tempInput;
                Total = Number;
                return tempOper1;
            }
            else
            {
                tempOper2 = tempInput;
                Total = Number;
                Number = null;
                return tempOper2;
            }
        }

        //Methods - calculations - requires 1 input number
        public double GetCalculator1NumResults(string operation)
        {
             switch (operation)
            {
                //Calc functions: x^2,sqR,cube,cubeR,abs
                case "x^2":
                    answer = Squared(double.Parse(Total));
                    tempOper1 = null;
                    Total = null;
                    Number = null;
                    break;
                case "sqR":
                    answer = SquareRoot(double.Parse(Total));
                    tempOper1 = null;
                    Total = null;
                    Number = null;
                    break;
                case "cube":
                    answer = Cubed(double.Parse(Total));
                    tempOper1 = null;
                    Total = null;
                    Number = null;
                    break;
                case "cubeR":
                    answer = CubeRoot(double.Parse(Total));
                    tempOper1 = null;
                    Total = null;
                    Number = null;
                    break;
                case "abs":
                    answer = Absolute(double.Parse(Total));
                    tempOper1 = null;
                    Total = null;
                    Number = null;
                    break;
                default:
                    break;
            }
            return answer;
         }
        //Methods - calculations - requires 2 input numbers
        public double GetCalculator2NumResults(string operation)
        {
            switch (operation)
            {
                //Calc functions: +,-,/,*,%
                case "+":
                    answer = Add(double.Parse(Total), double.Parse(Number));
                    tempOper2 = null;
                    Total = null;
                    Number = null;
                    break;
                case "-":
                    answer = Subtract(double.Parse(Total), double.Parse(Number));
                    tempOper2 = null;
                    Total = null;
                    Number = null;
                    break;
                case "/":
                    answer = Divide(double.Parse(Total), double.Parse(Number));
                    tempOper2 = null;
                    Total = null;
                    Number = null;
                    break;
                case "*":
                    answer = Multiply(double.Parse(Total), double.Parse(Number));
                    tempOper2 = null;
                    Total = null;
                    Number = null;
                    break;
                case "%":
                    answer = Remainder(double.Parse(Total), double.Parse(Number));
                    tempOper2 = null;
                    Total = null;
                    Number = null;
                    break;
                default:
                    break;
            }
            return answer;
        }           
            
        //Function Calculation Methods
        public double Add(double first, double second)
        {
            return first + second;
        }

        public double Subtract(double first, double second)
        {
            return first - second;
        }

        public double Divide(double first, double second)
        {
            if (second == 0)
            {
                return 0;
            }
            return first / second;
        }

        public double Multiply(double first, double second)
        {
            return first * second;
        }

        public double Remainder(double first, double second)
        {
            if (second == 0)
            {
                return 0;
            }
            return first % second;
        }

        public double Squared(double first)
        {
            return first * first;
        }

        public double SquareRoot(double first)
        {
            if (first < 0)
            {
                return 0;
            }
            return Math.Sqrt(first);
        }

        public double Cubed(double first)
        {
            return first * first * first;
        }
        public double CubeRoot(double first)
        {
            if (first <= 0)
            {
                return 0;
            }
            return Math.Pow(first, (double)1.0 / 3.0);
        }
        public double Absolute(double first)
        {
            return Math.Abs(first);
        }
        
        //Methods - validating whether input is a number or operator
            public bool IsValidNumber(string input)
        {
            double parse;
            return double.TryParse(input, out parse);
        }

        //validating if operation is in readonly array => return true
        public bool IsValidOperation1Num(string input) => operations1Num.Contains(input);
        public bool IsValidOperation2Num(string input) => operations2Num.Contains(input);
     
    }
}
